### PR TITLE
Use a generic wrapper for audit and metrics

### PR DESCRIPTION
**Pull request by dmage**

### CODE REVIEW

File: pkg/dockerregistry/server/wrapped/zz_funcname_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/docker/distribution/context"
+)
+
+func TestFuncName(t *testing.T) {
```
Comment (bparees) [inline pkg/dockerregistry/server/wrapped/zz_funcname_test.go]: would it make more sense to call this a wrapper test?

File: pkg/dockerregistry/server/wrapped/zz_wrapper.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package wrapped
```
Comment (bparees) [inline pkg/dockerregistry/server/wrapped/zz_wrapper.go]: why "zz_wrapper.go" and not just "wrapper.go"?

**Comment by miminar** [comment]
LGTM as long as Ben's questions are resolved.

File: pkg/dockerregistry/server/wrapped/zz_funcname_test.go (hunk 1 above)
Comment (dmage) [reply pkg/dockerregistry/server/wrapped/zz_funcname_test.go]: renamed

File: pkg/dockerregistry/server/wrapped/zz_wrapper.go (hunk 2 above)
Comment (dmage) [reply pkg/dockerregistry/server/wrapped/zz_wrapper.go]: renamed

**Comment by dmage** [comment]
@legionus done, the error middleware is splitted in two pieces.

**Pull request merged**